library/ssl_tls13_server.c

```diff
@@ -138,28 +139,36 @@ static int ssl_tls13_offered_psks_check_identity_match_ticket(
     }
     memcpy(ticket_buffer, identity, identity_len);
 
-    if ((ret = ssl->conf->f_ticket_parse(ssl->conf->p_ticket,
-                                         session,
-                                         ticket_buffer, identity_len)) != 0) {
-        if (ret == MBEDTLS_ERR_SSL_INVALID_MAC) {
-            MBEDTLS_SSL_DEBUG_MSG(3, ("ticket is not authentic"));
-        } else if (ret == MBEDTLS_ERR_SSL_SESSION_TICKET_EXPIRED) {
+    ret = ssl->conf->f_ticket_parse(ssl->conf->p_ticket,
+                                    session,
+                                    ticket_buffer, identity_len);
+    if (ret == 0) {
+        ret = SSL_TLS1_3_PSK_IDENTITY_MATCH;
+    } else {
+        if (ret == MBEDTLS_ERR_SSL_SESSION_TICKET_EXPIRED) {
             MBEDTLS_SSL_DEBUG_MSG(3, ("ticket is expired"));
+            ret = SSL_TLS1_3_PSK_IDENTITY_MATCH_BUT_PSK_NOT_USABLE;
         } else {
-            MBEDTLS_SSL_DEBUG_RET(1, "ticket_parse", ret);
+            if (ret == MBEDTLS_ERR_SSL_INVALID_MAC) {
+                MBEDTLS_SSL_DEBUG_MSG(3, ("ticket is not authentic"));
+            } else {
+                MBEDTLS_SSL_DEBUG_RET(1, "ticket_parse", ret);
+            }
+            ret = SSL_TLS1_3_PSK_IDENTITY_DOES_NOT_MATCH;
         }
     }
 
     /* We delete the temporary buffer */
     mbedtls_free(ticket_buffer);
 
-    if (ret == 0 && session->tls_version != MBEDTLS_SSL_VERSION_TLS1_3) {
-        MBEDTLS_SSL_DEBUG_MSG(3, ("Ticket TLS version is not 1.3."));
-        /* TODO: Define new return value for this case. */
-        ret = MBEDTLS_ERR_SSL_BAD_PROTOCOL_VERSION;
+    if (ret != SSL_TLS1_3_PSK_IDENTITY_MATCH) {
+        goto exit;
     }
 
-    if (ret != 0) {
+    ret = SSL_TLS1_3_PSK_IDENTITY_MATCH_BUT_PSK_NOT_USABLE;
```

Comment:
I prefer to keep it here rather that repeat it four times. I have added a comment. Factorizing this affectation does not probably make any difference for a suitably smart compiler but I have no idea if actually it would be just a few of them or all of them.